src/sgame/sg_entities.cpp

```diff
@@ -80,45 +80,43 @@ angles and bad trails.
 */
 gentity_t *G_NewEntity()
 {
-	int       i, force;
-	gentity_t *newEntity;
-
-	newEntity = nullptr; // shut up warning
-	i = 0; // shut up warning
+	// we iterate through all the entities and look for a free one that was allocated enough time ago,
+	// as well as one that died recently in case the first kind is not available
+	gentity_t *newEntity = &g_entities[ MAX_CLIENTS ];
+	gentity_t *forcedEnt = nullptr;
+	int i;
 
-	for ( force = 0; force < 2; force++ )
+	for ( i = MAX_CLIENTS; i < level.num_entities; i++, newEntity++ )
 	{
-		// if we go through all entities first and can't find a free one,
-		// then try again a second time, this time ignoring times
-		newEntity = &g_entities[ MAX_CLIENTS ];
-
-		for ( i = MAX_CLIENTS; i < level.num_entities; i++, newEntity++ )
+		if ( newEntity->inuse )
 		{
-			if ( newEntity->inuse )
-			{
-				continue;
-			}
-
-			// the first couple seconds of server time can involve a lot of
-			// freeing and allocating, so relax the replacement policy
-			if ( !force && newEntity->freetime > level.startTime + 2000 && level.time - newEntity->freetime < 1000 )
-			{
-				continue;
-			}
-
-			// reuse this slot
-			G_InitGentity( newEntity );
-			return newEntity;
+			continue;
 		}
 
-		if ( i != MAX_GENTITIES )
+		// the first couple seconds of server time can involve a lot of
+		// freeing and allocating, so relax the replacement policy
+		if ( newEntity->freetime > level.startTime + 2000 && level.time - newEntity->freetime < 1000 )
 		{
-			break;
+			// we could look for the oldest one, but who cares, really
+			forcedEnt = newEntity;
```

Comment:
Maybe assign to `forcedEnt` only if it is null? That way the lowest possible entity num is used, consistent with the non-forced case.

Reply:
I thought it was OK to avoid 3 more lines of code and not worry much about which entity to reuse in rare and extreme cases. But you're right really, it's cleaner to choose a slot with some logic rather than picking the last one.  
I would suggest to pick the slot that was freed the longest ago rather than the lowest one: this would reduce the amount of glitches players may see when slots are reused too quickly.

Let me know which policy you prefer and should I amend and force-push a single commit, or push this changes in a separate one?

By the way, a small note about this whole function: if it were ever to need further optimization, a much better approach would be to use a circular buffer of freed entity slots, instead of iterating through the entities array every time. I was unsure about changing this part of the code that much though.

Reply:
> I would suggest to pick the slot that was freed the longest ago rather than the lowest one: this would reduce the amount of glitches players may see when slots are reused too quickly.

OK sounds good

Reply:
Just pushed the change. This time I included some debug messages too (hidden behind `g_debugEntities`, like they did in `G_FreeEntity`).